frontend/src/metabase-lib/v1/expressions/formatter/formatter.ts

```diff
@@ -0,0 +1,474 @@
+
+function formatDimension(
+  path: AstPath<FieldReference>,
+  options: FormatOptions,
+): Doc {
+  const { query, stageIndex, expressionIndex } = options;
+
+  if (!query) {
+    throw new Error("`query` is a required parameter to format expressions");
+  }
+
+  const columns = Lib.expressionableColumns(query, stageIndex, expressionIndex);
+  const [columnIndex] = Lib.findColumnIndexesFromLegacyRefs(
+    query,
+    stageIndex,
+    columns,
+    [path.node],
+  );
+
+  const column = columns[columnIndex];
+  if (!column) {
+    return "";
+  }
+
+  const info = Lib.displayInfo(query, stageIndex, column);
+  return formatDimensionName(info.longDisplayName, options);
+}
+
+function formatMetric(path: AstPath<MetricAgg>, options: FormatOptions): Doc {
+  const [, metricId] = path.node;
+
+  const { query, stageIndex } = options;
+
+  if (!query) {
+    throw new Error("`query` is a required parameter to format expressions");
+  }
+
+  const metric = Lib.availableMetrics(query, stageIndex).find(metric => {
+    const [_type, availableMetricId] = Lib.legacyRef(query, stageIndex, metric);
+    return availableMetricId === metricId;
+  });
+
+  if (!metric) {
+    throw new Error(`metric with ID: ${metricId} does not exist`);
+  }
+
+  const displayInfo = Lib.displayInfo(query, stageIndex, metric);
+  return formatMetricName(displayInfo.displayName, options);
+}
+
+function formatSegment(path: AstPath<SegmentFilter>, options: FormatOptions) {
+  const { stageIndex, query } = options;
+
+  if (!query) {
+    throw new Error("`query` is a required parameter to format expressions");
+  }
+
+  const [, segmentId] = path.node;
+  const segment = Lib.availableSegments(query, stageIndex).find(segment => {
+    const [_type, availableSegmentId] = Lib.legacyRef(
+      query,
+      stageIndex,
+      segment,
+    );
+
+    return availableSegmentId === segmentId;
+  });
+
+  if (!segment) {
+    throw new Error(`segment with ID does not exist: ${segmentId}`);
+  }
+
+  const displayInfo = Lib.displayInfo(query, stageIndex, segment);
+  return formatSegmentName(displayInfo.displayName, options);
+}
+
+function formatFunctionOptions(options: CallOptions): Doc | null {
+  // HACK: very specific to some string/time functions for now
+  if (Object.prototype.hasOwnProperty.call(options, "case-sensitive")) {
+    const caseSensitive = options["case-sensitive"];
+    if (!caseSensitive) {
+      return formatStringLiteral("case-insensitive");
+    }
+  }
+  if (Object.prototype.hasOwnProperty.call(options, "include-current")) {
+    const includeCurrent = options["include-current"];
+    if (includeCurrent) {
+      return formatStringLiteral("include-current");
+    }
+  }
+  return null;
+}
+
+// Helper to recurse into an AST node that is not easily expressed as a
+// property of the node currently in path.
+// We will need this when switching to Lib.expressionParts, since the cljs
+// objects can't be traversed natively by prettier.
+function recurse<T, R>(
```

Comment:
@romeovs `Lib.ExpressionParts` are not CLJS objects. Only `column`s inside are, which should not be traversed into. I.e. this is how it looks like:
```
{
  operator: "+",
  args: [
    <column>,
   {
     operator: "-",
    args: [1, 2]
   }
  ]
}
```

Reply:
Good to know! That makes recursing into them even easier.

Reply:
And for `column`, there is `Lib.isColumnMetadata` type guard